### PR TITLE
MAFA 9

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Yet another run of micro optimizations:

* defer seeks and cache seek status
  This avoids repeated seeks to the beginning of the start/end bounds
  in RocksDB. Additionally, it delays seeking until the data is actually
  required.
* simplify the hash function for AqlValues to do less work
  The hash function is called very often, so saving instructions here
  should be at least slightly beneficial.
* ~~add a few hints to methods so that they can avoid repeated checks
  for whether a valid dataRow is available or not. This may save a few
  ms in huge queries.~~

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as *shell_server_aql*.

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10652/

### CODE REVIEW

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10669/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10727/

File: arangod/RocksDBEngine/RocksDBIterators.cpp [hunk 1]
```diff
@@ -87,7 +95,7 @@ bool RocksDBAllIndexIterator::nextImpl(LocalDocumentIdCallback const& cb, size_t
 
   TRI_ASSERT(limit > 0);
 
-  do {
```
Comment (graetzer) [inline arangod/RocksDBEngine/RocksDBIterators.cpp:90]: but limit > 0 for first iteration

Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBIterators.cpp:90]: I think using `do { ... } while (limit > 0)` and `while (limit > 0) { ... }` are equivalent here, as before the loop we always check for a limit value of 0.
Anyway, @goedderz requested the `do { ... } while (limit > 0)` stuff to be turned into `while (limit > 0) { ... }` in some other places, so I did this here as well just to get this PR finally through.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10743/

**Comment by jsteemann** [comment]
@graetzer : the goal is to not perform a Seek if it is not needed at all **and** to avoid redundant Seeks. I think even a simple query did multiple Seeks one after the other, which definitely is unnecessary if the iterator was not moved in between.

**Pull request merged**